Documentation/Contributing.md

```diff
@@ -0,0 +1,348 @@
+
+This command builds the Package Manager inside the `.build/` directory.
+    Run the bootstrap script to rebuild after making a change to the source
+    code.
+
+<a id="self-hosting">*D) _Self Hosting_*</a>:  
+
+It is possible to build SwiftPM with itself using SwiftPM present in latest
+release of Xcode or the latest trunk snapshot on Linux.
+
+```bash
+# Build:
+$> swift build
+
+# Run all tests.
+$> swift test --parallel
+
+# Run a single test.
+$> swift test --filter PackageGraphTests.DependencyResolverTests/testBasics
+
+# Run tests for the test targets BuildTests and WorkspaceTests, but skip some test cases.
+$> swift test --filter BuildTests --skip BuildPlanTests --filter WorkspaceTests --skip InitTests
+```
+
+Note: PackageDescription v4 is not available when developing using this method.
+
+This method can also used be used for performance testing. Use the following
+command run SwiftPM's performance tests:
+
+```bash
+$> export TSC_ENABLE_PERF_TESTS=1
+$> swift test -c release --filter PerformanceTests
+```
+
+<a id="swift-compiler-build-script">*E) _Using the Swift Compiler Build Script_*</a>:  
+
+Follow [these](https://github.com/apple/swift#getting-started) instructions to
+get the Swift sources and then execute the `build-script` using swiftpm preset:
+
+### macOS
+
+```bash
+$> ./swift/utils/build-script --preset=buildbot_swiftpm_macos_platform,tools=RA,stdlib=RA
+```
+
+### Linux
+
+```bash
+$> ./swift/utils/build-script --preset=buildbot_swiftpm_linux_platform,tools=RA,stdlib=RA
+```
+
+Once the build is complete, you should be able to run the swiftpm binaries from the build folder.
+
+### Example
+
+```bash
+$> cd /tmp && mkdir hello && cd hello
+$> /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-package init
+$> /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-build
+```
+
+8. Test the Swift Package Manager.
+
+```bash
+$> Utilities/bootstrap test
+```
+
+## Getting Started
+1. Pull the SwiftPM repository:
+```bash
+$> git clone https://github.com/apple/swift-package-manager.git
+```
+2. Run your first build:
+```bash
+$> cd swift-package-manager
+$> swift build
+```
+Make sure the build did not fail.  
+
+A `.build/` folder will be generated and it should have inside a similar structure (including build binaries):
+```bash
+artifacts/
+checkouts/
+debug/
+repositories
+x86_64-apple-macosx
+debug.yaml
+manifest.db
+workspace-state.json 
+```
+Binaries (in the example above) are in `x86_64-apple-macosx/`.  
+If you need to build the generated binaries, run `swift-build` in inside `.build/`:
+```bash
+./.build/x86_64-apple-macosx/debug/swift-build
+```
+
+## Running Tests
+### Run all tests.
+```bash
+$> swift test --parallel
+```
+### Run a single test.
+```bash
```

Comment:
Add short explanation above:

`For example, to only run the PackageGraphTests.DependencyResolverTests/testBasics`